#4 Support for tar/tar.gz archives

unicolet opened the pull request:
Fixes #4 

This PR picks up from where #5  left and: I merged with current master, then basically rewrote the test to decouple test outcome from tar on-disk storage.

@Zlika let me know your thoughts on this approach, thanks :-)

unicolet commented:
Adding IT

unicolet commented:
@Zlika should be ready for review, thanks :-)

Zlika commented on src/main/java/io/github/zlika/reproducible/TarStripper.java:
```diff
@@ -0,0 +1,151 @@
+        entry.setModTime(0);
+        // Remove GID and UID that are not deterministic, to evaluate though
+        entry.setGroupId(1000);
+        entry.setUserId(1000);
```
I think we should better set groupId and userId to 0 (i.e. root), because:
* we are sure that the "root" account exists
* according to what I have read, it seems that tar uses the current user's ID when extracting a file owned by root (to be checked)

Zlika commented on src/main/java/io/github/zlika/reproducible/UniversalTarConsumer.java:
```diff
@@ -0,0 +1,80 @@
+/**
+ * Process tar formats: tar, tar.gz, tar.bz2.
+ * @author Umberto Nicoletti (umberto.nicoletti@gmail.com)
+ * @version $Id$
```
Useless annotation.

Zlika commented on pom.xml:
```diff
@@ -9,7 +9,7 @@
   <description>This Plug-In allows to generate byte-for-byte reproducible artifacts</description>
   <url>http://zlika.github.io/reproducible-build-maven-plugin</url>
-  
+
```
Avoid committing files that only differ by whitespaces, that pollutes the diff.

unicolet replied:
@Zlika sorry about that, I was not sure if this came from previous PR. I'll be more careful in the future :-)

Zlika commented on src/it/strip-tar/assembly.xml:
```diff
@@ -0,0 +1,28 @@
+        <dependencySet>
+            <outputDirectory>/</outputDirectory>
+            <useProjectArtifact>true</useProjectArtifact>
+            <unpack>true</unpack>
```
If unpack=false (which is the default value) I don't think we will have a reproducible build because the maven-assembly-plugin:single goal is called in the package phase, whereas the stripping of the jar file is done in the next phase (to be sure it is called after the jar file is produced). So in this case you would include in the tarball a JAR file that is not reproducible, and then the tar file will be be reproducible.

Zlika commented on src/it/strip-tar/build1.properties:
```diff
@@ -0,0 +1,2 @@
+# current tar archive strip implementation does not 'fix' content under META-INF
+# user.name=user1
```
If we only put non-unpacked JAR files in the tar archive then we don't need to do anything special. Anyway, you should not comment this line, because you should have the same archive whatever your user name is. If not, the archive is not reproducible.

unicolet replied:
@Zlika AFAIU that goes in the manifest, which we don't clean up for tar archives. So, if we uncomment that line then we need to strip manifests as well or the IT will fail. I'm not an expert on the topic though so I'll follow your advice :-)

unicolet replied:
@Zlika I think I just understood what you meant here, sorry I'm proficient at mvn, but not as expert as you :-) Let me try to fix this

Zlika commented on src/main/java/io/github/zlika/reproducible/FileConsumer.java:
```diff
@@ -0,0 +1,32 @@
+ * @author Umberto Nicoletti (umberto.nicoletti@gmail.com)
+ * @version $Id$
+ */
+interface FileConsumer
```
Creating a new interface just because the "strip" method of interface Stripper throws IOException instead of MojoExecutionException seems overkill. Moreover, if you look at the implementation classes, there are in fact throwing an IOException that is catched to be replaced by a MojoExecutionException!
It seems that this new interface only add complexity for no obvious benefit.

unicolet replied:
@Zlika Thanks for the feedback and I see your point :-)

This refactoring was triggered by checkstyle complaining that the `strip()` method was too long and indeed it was too long and contained duplicated code. Initially I wanted to use a `Consumer` lambda, but then I could not throw any checked exception, so I took this slightly longer path.
The benefit of the interface is that it removes a some more duplicated code (not a lot, tbh). :-)

I could drop the interface, but I'd keep the 2 implementations and adapt `strip` and `process` accordingly. Wdyt?

Zlika commented on src/test/java/io/github/zlika/reproducible/ZipStripperTest.java:
```diff
@@ -28,23 +28,25 @@
-     * @throws IOException 
+     * 
+     * @throws IOException
+     *             in case of error on test file operations
```
Useless changeset that only differs on white spaces.
That's why I hate the IDE's automatic file formatters.

unicolet replied:
Apologies again :-)

unicolet commented:
@Zlika dropped the interface (I did not realize that `Stripper` interface and mine were so close, because I'm new to the code base :-)) and added IT to verify reproducible build for assembly plugin with `unpack=false` setting. Please have a look again, thanks:-)

unicolet commented:
@Zlika I think I addressed your remark in https://github.com/Zlika/reproducible-build-maven-plugin/pull/14#discussion_r167591219

Let me know your thoughts and thanks :-)

Zlika commented:
Hi Umberto. Thank you for your update.
To clean your PR, could you please revert the "white space only" changes? Also in the ZipStripper.java, a large part of the diff is just a move of some methods. Could you please try to minimize these unnecessary changes? That would make the PR and its merge cleaner.

Concerning the integration tests, I'm surprised they pass. I think you should update their pom.xml file this way:
* move the declaration of the reproducible-build-maven-plugin BEFORE the declaration of the maven-assembly-plugin (because when 2 plugins are used during the same phase, the first one activated is the first one declared in the pom)
* declare 2 executions for the strip-jar goal: one during the "package" phase (so that the produced artifact is stripped before it can be packaged by maven-assembly-plugin), and one during the "pre-integration-test" (just after the "package" phase) to strip the archive produced by maven-assembly-plugin.

Regards

Zlika commented:
Another remark: I don't think there is a need for 2 integration tests (for packed/unpacked) because that does not add any code coverage. I would only keep the "packed" version (but renaming  src/it/strip-tar-packed to  src/it/strip-tar) because this is the most common use case.
However, in the same integration test, I would add other assembly descriptors for tar.gz and tar.bz formats because that would improve code coverage.

unicolet commented:
@Zlika sure, I'll try to push the changes today. Thanks :-)

unicolet commented:
@Zlika in the recent push:

1. reverted formatting/white space changes
2. add IT for `.tar.gz` and `.tar.bz2`
3. dropped IT for unpacked

Please have a look, thanks!

Zlika commented:
Thank you.
Maybe I was not crystal clear for the IT: I would prefer a single IT folder (src/it/strip-tar) for this feature, but with several assembly files to produce and test all extensions (*.tar, *.tar.gz, *.tar.bz) in a single integration tests.

unicolet commented:
@Zlika my bad, apologies. I missed that point :-)

Zlika commented:
Thanks for your PR. I will take some time to test it and update the documentation before releasing the 0.5 version.

unicolet commented:
@Zlika thank you!

Zlika commented:
@unicolet Version 0.5 of the plugin has been released. Thank you for your contribution.

tglman commented:
Hi @Zlika, @unicolet,

Thank you for finish my work, I didn't have time to curry on this on my free time, but this is a really good news :)

Thank you again.

Zlika commented:
Thank you @tglman for your initial work. I credited both of you in the changelog for this feature.
